### PR TITLE
🔀 ♻️ Implement an interface dedicated for 'playing' through the stories

### PR DESCRIPTION
**Pull request by ReaganCn**
# Description

Created `BotEnginePlay` and implement an interface thats only responsible for playing through the stories. So that we can call `bot.play` in the Bot Engine Manager

## Type of change

- [x] Refactor

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203762063497913

### CODE REVIEW

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+    // Save the message
+    sideOperations.push(this.save(message, endUser.id));
+
+    // Reply To the end user
+    await this._sendBlockMessage(newStoryBlock, endUser.phoneNumber);
+
+    // Update the End User Position
+    const newPosition: EndUserPosition = { currentBlock: newStoryBlock };
+    sideOperations.push(this.__move(newPosition, endUser.id));
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:138]: Don't nest actions.

const moveToNext = this.__move...
sideOperations.push(moveToNext)

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    // Here is where the message chaining happens. 
+    //  If it is a text block we find the next block and send it. 
+    //    Our loop ends when we hit a story block type that is not specified here.
+    while (currentBlock.type === StoryBlockTypes.TextMessage || currentBlock.type == StoryBlockTypes.Image) {
+
+      // Get the next block in the story
+      currentBlock = await this.__getNextBlock(endUser, { currentBlock });
+
+      // Save the message
+      sideOperations.push(this._saveBlockAsMessage(currentBlock, endUser.id));
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:153]: Same, don't nest operations

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+      this._tools.Logger.log(() => `[EngineBotManager] - Next Block #${count} : ${JSON.stringify(currentBlock)}`);
+
+      // Inject variable to message
+      currentBlock.message = variableInjectorService.injectVariableToText(currentBlock.message, endUser);
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:158]: Naming could be better

mailMergeVariables

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+      // Send the message back to the end user
+      await this._sendBlockMessage(currentBlock, endUser.phoneNumber);
+
+      // Update the End User Position
+      sideOperations.push(this.__move({ currentBlock: currentBlock }, endUser.id));
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:164]: No nested actions

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+      count++;
+    }
+
+    return sideOperations;
+  }
+
+  /**
+   * Updates the position of the end user with the block we send back to them
+   */
+  private async __move(newPosition: EndUserPosition, endUserId: string): Promise<EndUserPosition>
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:175]: Should be batched in mid-term for performance

File: libs/functions/bot-engine/main/src/lib/services/bot-engine.main.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * The Base Class for the bot engine.
+ * 
+ * This class only contains methods that can be used throughout our engine, 
+ *    regardles of the chat status i.e. Running, Paused, Taking to Agent
+ * 
+ * This way we can have a class for the different status as they will be processed differently.
+ */
+export class BotEngineMain
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine.main.ts:22]: Duplication of the bot play?

File: libs/functions/bot-engine/main/src/lib/services/bot-engine.main.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  /**
+   * We need to keep the messages interchanged between the end user and the bot engine.
+   * 
+   * This method saves thoses messages and processes any media sent
+   */
+  public async save(message: Message, endUserId: string) 
+  {
+    if (message.type == MessageTypes.AUDIO || message.type == MessageTypes.VIDEO || message.type == MessageTypes.IMAGE) {
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/bot-engine.main.ts:55]: Should be done at Message parser level. You can have subtypes of MessageParser which is fileMessage, which does this

Then FileMessage is next inherited by the specific file types i.e. AudioMessage, ...

Comment (jrosseel) [reply libs/functions/bot-engine/main/src/lib/services/bot-engine.main.ts:55]: So save method in this subclass does
{
 fileMessage.url = ...
 return super.save(fileMessage, endUserId)
}

File: libs/functions/bot-engine/main/src/lib/services/chat-commands/chat-commands-manager.service.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+  defaultStory: string;
+
+  constructor(
+    private _endUserDataService$: EndUserDataService, 
+    private _activeChannel: ActiveChannel, 
+    private _processMessageService$: ProcessMessageService,
+    private _tools: HandlerTools) 
+  {
+    this.orgId =  _activeChannel.channel.orgId
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/chat-commands/chat-commands-manager.service.ts:22]: ; at the end

**Review by jrosseel — CHANGES_REQUESTED** [review]

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    sideOperations.push(this.__move(newPosition, endUser.id));
+
+    let count = 1;
+
+    let currentBlock = newStoryBlock;
+
+    // Here is where the message chaining happens. 
+    //  If it is a text block we find the next block and send it. 
+    //    Our loop ends when we hit a story block type that is not specified here.
+    while (currentBlock.type === StoryBlockTypes.TextMessage || currentBlock.type == StoryBlockTypes.Image) {
```
[the remark this answers is not in the document]
Comment (ReaganCn) [reply libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:150]: Storyblocks just have the property `type`, whose type is `StoryBlockTypes` and  blocks mainly set it to textmessage, image, video etc. There is also 'IO' and 'Input' in storyblock types in addition to the different block types.  Perhaps we add another property to storyblocks which shows whether its an input or output block?

File: libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts (hunk 5 above)
Comment (ReaganCn) [reply libs/functions/bot-engine/main/src/lib/services/bot-engine-play.service.ts:182]: It is pushed to sideOperations to be resolved later. Though an optimization of how we handle these types of operations can be done

**Pull request closed without merging**